### ISSUE TITLE
-audioadddev flag issue

### ISSUE THREAD

**Issue by Lucarda** [pure-data/pure-data#1557]
if I do `pd -asio -listdev` i get:

````
audio input devices:
1. MME: Microsoft Sound Mapper - Input
2. MME: Internal Microphone (Conexant 2
3. ASIO: ASIO4ALL v2
audio output devices:
1. MME: Microsoft Sound Mapper - Output
2. MME: Speakers (Conexant 20585 SmartA
3. ASIO: ASIO4ALL v2
API number 4

MIDI input devices:
1. loopMIDI Port
MIDI output devices:
1. Microsoft MIDI Mapper
2. Microsoft GS Wavetable Synth
3. loopMIDI Port
````

then if I do `pd -noprefs -asio -audioadddev "ASIO: ASIO4ALL v2"` i get:

````
Couldn't find audio input device: ASIO: ASIO4ALL v2
Couldn't find audio output device: ASIO: ASIO4ALL v2
````

Seems that, as audio is not initiated yet, `audio_nextsettings.a_api = API_DEFAULT;` in https://github.com/pure-data/pure-data/blob/master/src/s_audio.c#L206 takes effect for a short period and "ASIO: ASIO4ALL v2" is not found (it shouldn't as `API_DEFAULT` is `MMIO`).

If I change `s_audio.c#L206` to `audio_nextsettings.a_api = API_PORTAUDIO` Pd finds and uses "ASIO: ASIO4ALL v2".

PS: Tested on Windows but the issue might be affecting the other OSs as well.

**Comment by umlaeute** [comment]
so the problem is that it doesn't work if you add a device that is already there?

**Comment by Lucarda** [comment]
Yes.

**Comment by umlaeute** [comment]
so the question is: what is the desired behavior if you use `-audioadddev` for adding an already existing device?

**Comment by Spacechild1** [comment]
I think the key is the preceeding `-noprefs`

**Comment by umlaeute** [comment]
is it? the preferences do not *add* any devices (they are added by the audio backend).

or is the problem that `-audioadddev` tries to select a device?

i still would like to understand what the problem is *exactly*.

**Comment by Spacechild1** [comment]
> or is the problem that -audioadddev tries to select a device?

Yes. `-audioaddev` selects an audio device by name. You can use it repeatedly to select several devices.

The problem is that `-audioaddev` doesn't work if you start Pd with `-noprefs -asio`. I can confirm this!

FWIW, `-noprefs -asio -audiodev <n>` *does* work.

**Comment by Lucarda** [comment]
My use case for this flag had always been to start Pd with ASIO. My sound card is externally powered and is not always on.

I start Pd from 4 batch files. 2 of them are just `pd` (one for 32bit and the other for 64bit). The other 2 are :

````
%CD%/bin/pd -asio -audioadddev "ASIO: US-2x2 & US-4x4" -inchannels 4 -r 44100 -blocksize 64 -audiobuf 8
````
This way I don't have to change all settings manually on the "audio settings" dialog.

**Comment by Spacechild1** [comment]
The problem is the following: 

`-addaudiodev` calls `sys_audiodevnametonumber` which uses `audio_nextsettings.a_api`. However, `-asio` only sets the API in the local `t_audiosettings` instance. Since we call Pd with `-noprefs`, `audio_nextsettings` still contains default values (here: MMIO). Consequently,  `sys_audiodevnametonumber` can't find the audio device because it uses the wrong API.

I guess flags like `-asio`, `-mmio`, `-jack`, etc. could set the API in both `t_audiosettings` instances, but I haven't really thought it through yet.

IMO, a better solution would be that `sys_audiodevnametonumber` and `sys_audionumbertodevname` explicitly take the API as an argument. However, these functions are declared `EXTERN`, which means that changing the signature would be a breaking change. Alternatively, we could add private versions with the additional API argument and the existing (semi)public versions would merely call the private ones with `audio_nextsettings.a_api`.

**Comment by umlaeute** [comment]
so the problem is a regression?

(i obviously hardly ever use `-audioaddev`; in ALSA-land it can be used to *add* devices (hence the name!) that do not show up automatically in Pd's device-list (as Pd only lists "hardware devices" but not "software devices" (see also #1487); but i mostly use JACK (where this makes no sense anyhow), and ALSA only in reduced setups, where i never want to use the "software devices")


> However, these functions are declared `EXTERN`, which means that

Pd deals *very sloppily* with exported and private symbols. i think that most symbols exported via `EXTERN` are simply declared as public because of copy-and-paste (or because surrounding symbols in the header already had the `EXTERN` symbol).
it would be good to make a great symbol purge...

**Comment by Lucarda** [comment]
@spacechild1

> can't find the audio device because it uses the wrong API

Seems logical but it can't open a device even at the same (and not wrong) API. 

@umlaeute 

> so the problem is a regression?

No. I had just tested down to 0.48-1 an it never worked. IIRC I used this command mostly when testing AOO a couple of years ago but it seems I had to open audio settings just to select the sound card (the rest of the settings where ok).

It seems I hit an old bug.


PS: I can perfectly (and probably better) live with `pd -prefsfile ../asio`.

**Comment by Lucarda** [comment]
> It seems I hit an old bug.

No. Probably there's no bug at all and is just me misunderstanding the flag. If this is the case sorry for the noise and we close this one.

**Comment by umlaeute** [comment]
> Probably there's no bug at all and is just me misunderstanding the flag. If this is the case sorry for the noise and we close this one.

that's what i was wondering. hence my dumb questions to clarify what the actual bug was.

i think there is still room for improvement (apart from the documentation): e.g. plainly refusing to add devices that already exist (or discarding them with a message).
i don't know of the implications (e.g. whether there are setups that require multiple devices with the same name).


it would also help if we could clarify what the expectations are (and whether we need *another* flag for them)

**Comment by Spacechild1** [comment]
> (i obviously hardly ever use -audioaddev; in ALSA-land it can be used to add devices (hence the name!) that do not show up automatically in Pd's device-list (as Pd only lists "hardware devices" but not "software devices"

I think you're confusing it with `-alsaadd <name>` (which calls `alsa_adddev`).

`-audioaddev` just tries to find a device with a matching name in the device list and then adds it to the devices in the local `t_audiosettings` object.

---

Generally, `-audioadddev` is the counterpart to `-audiodev`. The difference is the following:
* `-audiodev` works with device *numbers* and can set one ore more devices in one go (it's possible to use comma separated values, like `-audiodev "0,1"`).
* `-audioaddev` works with device *names* and only allows to add a single device.

The problem with `-audioadddev` is that it *adds* a device, which makes it awkward to *set* a device (unless no audio preferences have been saved). @Lucarda basically tried to workaround this by adding `-noprefs` in the beginning. Apart from the fact that there is a bug with `-noprefs` + `-addaudiodev`, it is also not a good solution because `-nopref` ignores *all* core preferences - which is not really what we want here!

BTW, we have the same problem with `-mididev` vs `-midiadddev`.

One solution could be to add a flag that clears the current device list, e.g. `-audioclear`. You would use it like this:
`pd -audioclear -audioadddev "ASIO: ASIO4ALL v2"`

A better solution would be to have something like `-audiodev` that works with device *names* instead of *numbers*. Here's a not-so-polite solution: `-audiodev` could accept device names and if a device name contains a comma, it is bashed to some replacement symbol.

**Comment by umlaeute** [comment]
> I think you're confusing it with `-alsaadd <name>` (which calls alsa_adddev).

DoH. yes. :facepalm: 

> One solution could be to add a flag that clears the current device list, e.g. `-audioclear`.

`-audiocleardev` if you want.
`-audioclear` would just turn on crystal sound processing.

> The problem with `-audioadddev` is that it *adds* a device, but most of the time we really want to *set* devices.

indeed.
i do not think that a flag named `-audioadddev` should try to set the device.
if we want that we need an other flag.
i agree that `-audiodev` seems to be a likely candidate.

> Here's a not-so-polite solution:

?

>  `-audiodev` could also accept device names 

sounds reasonable.

> and if a device name contains a comma, it is bashed to some replacement character.

the typical way would be to use an escape character like `\`.

**Comment by Spacechild1** [comment]
> the typical way would be to use an escape character like \.

that would be the polite solution :-)
